qiskit/algorithms/optimizers/steppable_optimizer.py

```diff
@@ -0,0 +1,286 @@
+        specific knowledge about his function can catch this errors and handle them before passing
+        the result to the optimizer.
+
+        .. code-block:: python
+
+            import random
+            import numpy as np
+            from qiskit.algorithms.optimizers import GradientDescent
+
+            def objective(x):
+                if random.choice([True, False]):
+                    return None
+                else:
+                    return (np.linalg.norm(x) - 1) ** 2
+
+            def grad(x):
+                if random.choice([True, False]):
+                    return None
+                else:
+                    return 2 * (np.linalg.norm(x) - 1) * x / np.linalg.norm(x)
+
+
+            initial_point = np.random.normal(0, 1, size=(100,))
+
+            optimizer = GradientDescent(maxiter=20)
+            optimizer.initialize(x0=initial_point, fun=objective, jac=grad)
+
+            for _ in range(20):
+                ask_object = optimizer.ask()
+                evaluated_gradient = None
+
+                while evaluated_gradient is None:
+                    evaluated_gradient = grad(ask_object.x_center)
+                    optimizer._state.njev += 1
+
+                optmizer._state.nit += 1
+
+                tell_object = TellObject(eval_jac=evaluated_gradient)
+                optimizer.tell(ask_object=ask_object, tell_object=tell_object)
+
+            result = optimizer.create_result()
+
+
+
+    Users that aren't dealing with complicated functions and who are more familiar with step by step
+    optimization algorithms can use the :meth:`~.step` method which wraps the :meth:`~.ask`
+    and :meth:`~.tell` methods. In the same spirit the method :meth:`~.minimize` will optimize the
+    function and return the result without directly.
+
+    To see other libraries that use this interface one can visit:
+    https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html
+
+
+    """
+
+    def __init__(
+        self,
+        maxiter: int = 1000,
+    ):
+        """
+        Args:
+            maxiter: Number of steps in the optimization process before ending the loop.
+        """
+        super().__init__()
+        self._state: OptimizerState = None
+        self.maxiter = maxiter
+
+    def ask(self) -> AskObject:
+        """Ask the optimizer for a set of points to evaluate.
+
+        This method asks the optimizer which are the next points to evaluate.
+        These points can, e.g., correspond to function values and/or its derivative.
+        It may also correspond to variables that let the user infer which points to evaluate.
+        It is the first method inside of a :meth:`~.step` in the optimization process.
+
+        Returns:
+            Since the way to evaluate the function can vary much with different
+            optimization algorithms, the object will be a custom dataclass for each optimizer.
```

Comment:
Maybe you can move this more into the descriptive text and leave the text here simply stating what is returned. Either way you always expect these to be a subclass of an Ask right, which is what is meant by custom here.